packages/toolpad-app/src/toolpadDataSources/rest/client.tsx

```diff
@@ -201,6 +211,27 @@ function QueryEditor({
     }));
   }, []);
 
+  const handleBodyChange = React.useCallback((newBody: Maybe<Body>) => {
```

Comment:
should we have a reusable function for these? the callbacks seem pretty similar.

Reply:
but if we predict these might changeover time it's best to keep them separate though, just a thought

Reply:
One could invent a hook like:

```tsx
function useQueryChangeHandler<F extends Function>(
  setInput: React.Dispatch<React.SetStateAction<QueryEditorModel<FetchQuery>>>,
  property: string,
  producer: F,
): F {
  return React.useCallback<F>(
    (...args: any[]) => {
      setInput((existing) => ({
        ...existing,
        query: { ...existing.query, [property]: producer(...args) },
      }));
    },
    [producer, property, setInput],
  );
}
```

and use as 

```tsx
const handleMethodChange = useQueryChangeHandler('method', (event) => event.target.value);
const handleMethodChange = useQueryChangeHandler('url', (newUrl) => newUrl);
// ...
```

But that `produce` parameter would invalidate the callback on each render. So we could declare those functions in module scope, or wrap them with `useCallback` as well, but at which point do all the abstractions become less readable again? The way it's done now is idiomatic and explicit and easy to understand. I kind of prefer it, even though it could probably be expressed in less lines of code. I was thinking for the future [we could use `immer`](https://github.com/mui/mui-toolpad/issues/763) to remove some of the duplication inside the `setInput`.

e.g. from
```tsx
  const handleMethodChange = React.useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
    setInput((existing) => ({
      ...existing,
      query: { ...existing.query, method: event.target.value },
    }));
  }, []);
```
to
```tsx
  const handleMethodChange = React.useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) =>
      setInput(
        produce((draft) => {
          draft.query.method = event.target.value;
        }),
      ),
    [],
  );
```

Reply:
yeah, it's probably not worth introducing something so complex yet at least (we could not use a `produce` function/argument, but then we would probably be making the callback less flexible)
it was just a thought, i think we can keep this as it is it's pretty easy to read and understand.
i agree with using `immer` later for this kind of operations later though, makes them a bit cleaner!